File: internal/service/repair_service.go
package service

import (
	"fmt"
	"time"

	"github.com/hafizd-kurniawan/pos-baru/internal/domain/models"
	"github.com/hafizd-kurniawan/pos-baru/internal/repository"
)

type RepairService interface {
	CreateRepairOrder(request *models.RepairOrderCreateRequest, assignedBy int) (*models.RepairOrder, error)
	GetRepairOrder(id int) (*models.RepairOrder, error)
	GetRepairOrderByCode(code string) (*models.RepairOrder, error)
	ListRepairOrders(filter models.RepairOrderFilter, page, limit int) ([]models.RepairOrder, int, error)
	UpdateRepairOrder(id int, request *models.RepairOrderUpdateRequest) error
	UpdateRepairProgress(id int, request *models.RepairProgressUpdateRequest) error
	DeleteRepairOrder(id int) error

	// Spare parts management
	AddSparePartToRepair(repairID int, request *models.RepairSparePartCreateRequest) error
	RemoveSparePartFromRepair(repairID int, sparePartID int) error
	GetRepairSpareParts(repairID int) ([]models.RepairSparePart, error)

	// Statistics and reporting
	GetRepairStats(mechanicID *int, dateFrom, dateTo *time.Time) (map[string]interface{}, error)
	GetMechanicWorkload() ([]map[string]interface{}, error)

	// Vehicle management
	GetVehiclesNeedingRepairOrders() ([]models.Vehicle, error)
}

type repairService struct {
	repairRepo    repository.RepairRepository
	vehicleRepo   repository.VehicleRepository
	userRepo      repository.UserRepository
	sparePartRepo repository.SparePartRepository
}

func NewRepairService(repairRepo repository.RepairRepository, vehicleRepo repository.VehicleRepository, userRepo repository.UserRepository, sparePartRepo repository.SparePartRepository) RepairService {
	return &repairService{
		repairRepo:    repairRepo,
		vehicleRepo:   vehicleRepo,
		userRepo:      userRepo,
		sparePartRepo: sparePartRepo,
	}
}

func (s *repairService) CreateRepairOrder(request *models.RepairOrderCreateRequest, assignedBy int) (*models.RepairOrder, error) {
	// Validate vehicle exists and status
	vehicle, err := s.vehicleRepo.GetByID(request.VehicleID)
	if err != nil {
		return nil, fmt.Errorf("vehicle not found: %v", err)
	}

	// Check if vehicle can be repaired
	if vehicle.Status == models.VehicleStatusSold {
		return nil, fmt.Errorf("cannot repair sold vehicle")
	}

	// Validate mechanic exists and has correct role
	_, err = s.userRepo.GetByID(request.MechanicID)
	if err != nil {
		return nil, fmt.Errorf("mechanic not found: %v", err)
	}

	// if mechanic.RoleID != 3 { // Assuming role_id 3 is mechanic
	// 	return nil, fmt.Errorf("assigned user is not a mechanic")
	// }

	// Generate repair order code
	code := s.generateRepairCode()

	// Create repair order
	repair := &models.RepairOrder{
		Code:          code,
		VehicleID:     request.VehicleID,
		MechanicID:    request.MechanicID,
		AssignedBy:    assignedBy,
		Description:   request.Description,
		EstimatedCost: request.EstimatedCost,
		Status:        models.RepairStatusPending,
		Notes:         request.Notes,
	}

	err = s.repairRepo.Create(repair)
	if err != nil {
		return nil, fmt.Errorf("failed to create repair order: %v", err)
	}

	// Update vehicle status to in_repair
	err = s.vehicleRepo.UpdateStatus(request.VehicleID, string(models.VehicleStatusInRepair))
	if err != nil {
		return nil, fmt.Errorf("failed to update vehicle status: %v", err)
	}

	// Get complete repair order with relationships
	return s.repairRepo.GetByID(repair.ID)
}

func (s *repairService) GetRepairOrder(id int) (*models.RepairOrder, error) {
	return s.repairRepo.GetByID(id)
}

func (s *repairService) GetRepairOrderByCode(code string) (*models.RepairOrder, error) {
	return s.repairRepo.GetByCode(code)
}

func (s *repairService) ListRepairOrders(filter models.RepairOrderFilter, page, limit int) ([]models.RepairOrder, int, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 || limit > 100 {
		limit = 10
	}

	return s.repairRepo.List(filter, page, limit)
}

func (s *repairService) UpdateRepairOrder(id int, request *models.RepairOrderUpdateRequest) error {
	// Check if repair order exists
	repair, err := s.repairRepo.GetByID(id)
	if err != nil {
		return fmt.Errorf("repair order not found: %v", err)
	}

	// Validate status transition if status is being updated
	if request.Status != nil {
		err = s.validateStatusTransition(repair.Status, *request.Status)
		if err != nil {
			return err
		}
	}

	return s.repairRepo.Update(id, request)
}

func (s *repairService) UpdateRepairProgress(id int, request *models.RepairProgressUpdateRequest) error {
	// Check if repair order exists
	repair, err := s.repairRepo.GetByID(id)
	if err != nil {
		return fmt.Errorf("repair order not found: %v", err)
	}

	fmt.Printf("UpdateRepairProgress service: repair found ID=%d, current status=%s\n", repair.ID, repair.Status)

	// Validate status transition
	err = s.validateStatusTransition(repair.Status, request.Status)
	if err != nil {
		return err
	}

	fmt.Printf("UpdateRepairProgress service: updating to status=%s\n", request.Status)

	// Update repair progress
	err = s.repairRepo.UpdateProgress(id, request)
	if err != nil {
		return fmt.Errorf("failed to update repair progress: %v", err)
	}

	fmt.Printf("UpdateRepairProgress service: repair progress updated successfully\n")

	// Update vehicle status based on repair status
	var vehicleStatus models.VehicleStatus
	switch request.Status {
	case models.RepairStatusCompleted:
		vehicleStatus = models.VehicleStatusAvailable

		// Always calculate repair cost from spare parts used only
		spareParts, err := s.repairRepo.GetSpareParts(id)
		if err != nil {
			return fmt.Errorf("failed to get spare parts for cost calculation: %v", err)
		}

		totalSparePartsCost := 0.0
		for _, part := range spareParts {
			totalSparePartsCost += part.UnitPrice * float64(part.QuantityUsed)
		}

		// Update vehicle repair cost with spare parts total only
		err = s.vehicleRepo.UpdateRepairCost(repair.VehicleID, totalSparePartsCost)
		if err != nil {
			return fmt.Errorf("failed to update vehicle repair cost: %v", err)
		}

		fmt.Printf("UpdateRepairProgress service: vehicle repair cost updated to %.2f (from spare parts only)\n", totalSparePartsCost)

	case models.RepairStatusCancelled:
		vehicleStatus = models.VehicleStatusAvailable

	case models.RepairStatusInProgress:
		vehicleStatus = models.VehicleStatusInRepair

	default:
		// No vehicle status update needed for pending
		return nil
	}

	err = s.vehicleRepo.UpdateStatus(repair.VehicleID, string(vehicleStatus))
	if err != nil {
		return fmt.Errorf("failed to update vehicle status: %v", err)
	}

	fmt.Printf("UpdateRepairProgress service: vehicle status updated to %s\n", vehicleStatus)

	return nil
}

func (s *repairService) DeleteRepairOrder(id int) error {
	// Check if repair order exists
	repair, err := s.repairRepo.GetByID(id)
	if err != nil {
		return fmt.Errorf("repair order not found: %v", err)
	}

	// Only allow deletion if repair is pending or cancelled
	if repair.Status != models.RepairStatusPending && repair.Status != models.RepairStatusCancelled {
		return fmt.Errorf("cannot delete repair order in %s status", repair.Status)
	}

	err = s.repairRepo.Delete(id)
	if err != nil {
		return fmt.Errorf("failed to delete repair order: %v", err)
	}

	// Update vehicle status back to available if it was in repair
	if repair.Status == models.RepairStatusPending {
		err = s.vehicleRepo.UpdateStatus(repair.VehicleID, string(models.VehicleStatusAvailable))
		if err != nil {
			return fmt.Errorf("failed to update vehicle status: %v", err)
		}
	}

	return nil
}

func (s *repairService) AddSparePartToRepair(repairID int, request *models.RepairSparePartCreateRequest) error {
	// Check if repair order exists
	_, err := s.repairRepo.GetByID(repairID)
	if err != nil {
		return fmt.Errorf("repair order not found: %v", err)
	}

	// Check if spare part exists
	_, err = s.sparePartRepo.GetByID(request.SparePartID)
	if err != nil {
		return fmt.Errorf("spare part not found: %v", err)
	}

	return s.repairRepo.AddSparePart(repairID, request)
}

func (s *repairService) RemoveSparePartFromRepair(repairID int, sparePartID int) error {
	// Check if repair order exists
	_, err := s.repairRepo.GetByID(repairID)
	if err != nil {
		return fmt.Errorf("repair order not found: %v", err)
	}

	return s.repairRepo.RemoveSparePart(repairID, sparePartID)
}

func (s *repairService) GetRepairSpareParts(repairID int) ([]models.RepairSparePart, error) {
	// Check if repair order exists
	_, err := s.repairRepo.GetByID(repairID)
	if err != nil {
		return nil, fmt.Errorf("repair order not found: %v", err)
	}

	return s.repairRepo.GetSpareParts(repairID)
}

func (s *repairService) GetRepairStats(mechanicID *int, dateFrom, dateTo *time.Time) (map[string]interface{}, error) {
	return s.repairRepo.GetRepairStats(mechanicID, dateFrom, dateTo)
}

func (s *repairService) GetMechanicWorkload() ([]map[string]interface{}, error) {
	// This would need a separate query to get mechanics and their current workload
	// For now, return empty slice
	return []map[string]interface{}{}, nil
}

func (s *repairService) GetVehiclesNeedingRepairOrders() ([]models.Vehicle, error) {
	// Get all vehicles with in_repair status
	inRepairStatus := models.VehicleStatusInRepair
	vehicles, _, err := s.vehicleRepo.List(1, 100, &inRepairStatus) // Get up to 100 vehicles

	if err != nil {
		return nil, fmt.Errorf("failed to get vehicles: %v", err)
	}

	var vehiclesNeedingOrders []models.Vehicle

	// Check each vehicle to see if it has an active repair order
	for _, vehicle := range vehicles {
		// Check if vehicle has an active repair order (pending or in_progress)
		repairs, _, err := s.repairRepo.List(models.RepairOrderFilter{
			VehicleID: vehicle.ID,
			Status:    "", // Get all statuses
		}, 1, 10)

		if err != nil {
			fmt.Printf("Error checking repairs for vehicle %d: %v\n", vehicle.ID, err)
			continue
		}

		// Check if there's any active repair order (pending or in_progress)
		hasActiveRepair := false
		for _, repair := range repairs {
			if repair.Status == models.RepairStatusPending || repair.Status == models.RepairStatusInProgress {
				hasActiveRepair = true
				break
			}
		}

		// If no active repair order, add to list
		if !hasActiveRepair {
			vehiclesNeedingOrders = append(vehiclesNeedingOrders, vehicle)
		}
	}

	return vehiclesNeedingOrders, nil
}

// Helper methods

func (s *repairService) generateRepairCode() string {
	// Generate repair code in format RPR-YYYYMMDD-XXX
	now := time.Now()
	dateStr := now.Format("20060102")

	// This is a simple implementation. In production, you might want to:
	// 1. Query for the last repair order of the day
	// 2. Increment the sequence number
	// 3. Handle concurrent requests properly

	return fmt.Sprintf("RPR-%s-%03d", dateStr, now.Nanosecond()%1000)
}

func (s *repairService) validateStatusTransition(currentStatus, newStatus models.RepairStatus) error {
	// Define allowed status transitions
	allowedTransitions := map[models.RepairStatus][]models.RepairStatus{
		models.RepairStatusPending: {
			models.RepairStatusInProgress,
			models.RepairStatusCancelled,
		},
		models.RepairStatusInProgress: {
			models.RepairStatusCompleted,
			models.RepairStatusCancelled,
		},
		models.RepairStatusCompleted: {
			// No transitions allowed from completed
		},
		models.RepairStatusCancelled: {
			models.RepairStatusPending, // Allow reactivation
		},
	}

	allowed := allowedTransitions[currentStatus]
	for _, status := range allowed {
		if status == newStatus {
			return nil
		}
	}

	return fmt.Errorf("invalid status transition from %s to %s", currentStatus, newStatus)
}

File: pos_flutter_app/lib/features/repairs/presentation/services/repair_service.dart
import 'package:dio/dio.dart';

import '../../../../core/constants/api_endpoints.dart';
import '../../../../core/models/repair.dart';
import '../../../../core/network/api_client.dart';

class RepairService {
  final ApiClient _apiClient;

  RepairService({ApiClient? apiClient}) : _apiClient = apiClient ?? ApiClient();

  /// Get repairs with pagination
  Future<List<Repair>> getRepairs({
    required String token,
    String? status,
    int page = 1,
    int limit = 20,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'page': page.toString(),
        'limit': limit.toString(),
      };

      if (status != null) {
        queryParams['status'] = status;
      }

      final response = await _apiClient.get(
        ApiEndpoints.repairs,
        queryParameters: queryParams,
      );

      if (response.data['success'] == true) {
        final List<dynamic> repairsData =
            response.data['data']['repairs'] ?? [];
        return repairsData.map((data) => Repair.fromJson(data)).toList();
      } else {
        throw Exception(response.data['message'] ?? 'Failed to get repairs');
      }
    } catch (e) {
      throw Exception('Failed to get repairs: $e');
    }
  }

  /// Get repair by ID
  Future<Repair> getRepairById({
    required int id,
    required String token,
  }) async {
    try {
      final response = await _apiClient.get(
        ApiEndpoints.repairById(id),
      );

      if (response.data['success'] == true) {
        return Repair.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'Failed to get repair');
      }
    } catch (e) {
      throw Exception('Failed to get repair: $e');
    }
  }

  /// Create new repair
  Future<Repair> createRepair({
    required CreateRepairRequest request,
    required String token,
  }) async {
    try {
      final response = await _apiClient.post(
        ApiEndpoints.repairs,
        data: request.toJson(),
      );

      if (response.data['success'] == true) {
        return Repair.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'Failed to create repair');
      }
    } catch (e) {
      throw Exception('Failed to create repair: $e');
    }
  }

  /// Update repair
  Future<Repair> updateRepair({
    required int id,
    required UpdateRepairRequest request,
    required String token,
  }) async {
    try {
      final response = await _apiClient.put(
        ApiEndpoints.repairById(id),
        data: request.toJson(),
      );

      if (response.data['success'] == true) {
        return Repair.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'Failed to update repair');
      }
    } catch (e) {
      throw Exception('Failed to update repair: $e');
    }
  }

  /// Delete repair
  Future<void> deleteRepair({
    required int id,
    required String token,
  }) async {
    try {
      final response = await _apiClient.delete(
        ApiEndpoints.repairById(id),
      );

      if (response.data['success'] != true) {
        throw Exception(response.data['message'] ?? 'Failed to delete repair');
      }
    } catch (e) {
      throw Exception('Failed to delete repair: $e');
    }
  }

  /// Update repair progress
  Future<void> updateRepairProgress({
    required int id,
    required String status,
    required double actualCost,
    String? notes,
    required List<RepairSparePartRequest> spareParts,
    required String token,
  }) async {
    try {
      final response = await _apiClient.patch(
        ApiEndpoints.updateRepairProgress(id),
        data: {
          'status': status,
          'actual_cost': actualCost,
          'notes': notes,
          'spare_parts': spareParts.map((sp) => sp.toJson()).toList(),
        },
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      if (response.data['success'] != true) {
        throw Exception(
            response.data['message'] ?? 'Failed to update repair progress');
      }
    } catch (e) {
      throw Exception('Failed to update repair progress: $e');
    }
  }

  /// Add spare part to repair order
  Future<void> addSparePartToRepair({
    required int repairId,
    required int sparePartId,
    required int quantityUsed,
    required String token,
  }) async {
    try {
      final response = await _apiClient.post(
        ApiEndpoints.repairSpareParts(repairId),
        data: {
          'spare_part_id': sparePartId,
          'quantity_used': quantityUsed,
        },
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      if (response.data['success'] != true) {
        throw Exception(
            response.data['message'] ?? 'Failed to add spare part to repair');
      }
    } catch (e) {
      throw Exception('Failed to add spare part to repair: $e');
    }
  }

  /// Create new repair order
  Future<Repair> createRepairOrder({
    required String code,
    required int vehicleId,
    required int mechanicId,
    String? description,
    double estimatedCost = 0.0,
    String? notes,
    required String token,
  }) async {
    try {
      final response = await _apiClient.post(
        ApiEndpoints.repairs,
        data: {
          'code': code,
          'vehicle_id': vehicleId,
          'mechanic_id': mechanicId,
          'description':
              description ?? 'Repair created from mechanic spare parts page',
          'estimated_cost': estimatedCost,
          'notes': notes,
        },
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      if (response.data['success'] == true) {
        return Repair.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'Failed to create repair');
      }
    } catch (e) {
      throw Exception('Failed to create repair: $e');
    }
  }
}

/// Request model for repair spare part
class RepairSparePartRequest {
  final int sparePartId;
  final int quantityUsed;

  RepairSparePartRequest({
    required this.sparePartId,
    required this.quantityUsed,
  });

  Map<String, dynamic> toJson() {
    return {
      'spare_part_id': sparePartId,
      'quantity_used': quantityUsed,
    };
  }
}
